plugins/inputs/docker_log/docker_logs.go

```diff
@@ -0,0 +1,472 @@
+	}
+	return false
+}
+
+func (d *DockerLogs) stopAllReaders() error {
+	for _, container := range d.containerList {
+		container.Close()
+	}
+	return nil
+}
+
+func (d *DockerLogs) containerListUpdate(acc telegraf.Accumulator) error {
+	ctx, cancel := context.WithTimeout(context.Background(), d.Timeout.Duration)
+	defer cancel()
+	if d.client == nil {
+		return errors.New(fmt.Sprintf("%s : Dock client is null", ERR_PREFIX))
+	}
+	containers, err := d.client.ContainerList(ctx, d.opts)
+	if err != nil {
+		return err
+	}
+	for _, container := range containers {
+		if d.containerInContainerList(container.ID) {
+			continue
+		}
+		d.wg.Add(1)
+		/*Start a new goroutine for every new container that has logs to collect*/
+		go func(c types.Container) {
+			defer d.wg.Done()
+			logOptions := types.ContainerLogsOptions{
+				ShowStdout: true,
+				ShowStderr: true,
+				Timestamps: false,
+				Details:    true,
+				Follow:     true,
+				Tail:       "0",
+			}
+			logReader, err := d.client.ContainerLogs(context.Background(), c.ID, logOptions)
+			if err != nil {
+				acc.AddError(err)
+				return
+			}
+			d.addToContainerList(c.ID, logReader)
+			err = d.tailContainerLogs(c, logReader, acc)
+			if err != nil {
+				acc.AddError(err)
+			}
+			d.removeFromContainerList(c.ID)
+			return
+		}(container)
+	}
+	return nil
+}
+
+func (d *DockerLogs) tailContainerLogs(
+	container types.Container, logReader io.ReadCloser,
+	acc telegraf.Accumulator,
+) error {
+	c, err := d.client.ContainerInspect(context.Background(), container.ID)
+	if err != nil {
+		return err
+	}
+	/* Parse container name */
+	var cname string
+	for _, name := range container.Names {
+		trimmedName := strings.TrimPrefix(name, "/")
+		match := d.containerFilter.Match(trimmedName)
+		if match {
+			cname = trimmedName
+			break
+		}
+	}
+
+	if cname == "" {
+		return errors.New(fmt.Sprintf("%s : container name is null", ERR_PREFIX))
+	}
+	imageName, imageVersion := parseImage(container.Image)
+	tags := map[string]string{
+		"container_name":    cname,
+		"container_image":   imageName,
+		"container_version": imageVersion,
+	}
+	fields := map[string]interface{}{}
+	fields["container_id"] = container.ID
+	// Add labels to tags
+	for k, label := range container.Labels {
+		if d.labelFilter.Match(k) {
+			tags[k] = label
+		}
+	}
+	if c.Config.Tty {
+		err = pushTtyLogs(acc, tags, fields, logReader)
+	} else {
+		_, err = pushLogs(acc, tags, fields, logReader)
+	}
+	if err != nil {
+		return err
+	}
+	return nil
+}
+func pushTtyLogs(acc telegraf.Accumulator, tags map[string]string, fields map[string]interface{}, src io.Reader) (err error) {
+	tags["logType"] = "unknown" //in tty mode we wont be able to differentiate b/w stdout and stderr hence unknown
+	data := make([]byte, logBytesMax)
+	for {
+		num, err := src.Read(data)
+		if num > 0 {
+			fields["message"] = data[1:num]
+			acc.AddFields("docker_log", fields, tags)
+		}
+		if err == io.EOF {
+			fields["message"] = data[1:num]
```

Comment:
During testing I ran into a crash here, line numbers are a bit off because I had added some print statements, the panic actually occurred on line 273.
```
panic: runtime error: slice bounds out of range

goroutine 24 [running]:
github.com/influxdata/telegraf/plugins/inputs/docker_log.pushTtyLogs(0x2a03680, 0xc000140640, 0xc0005956b0, 0xc0005956e0, 0x29a5ae0, 0xc0000a6c80, 0x45248c0, 0x0)
        /home/dbn/src/influx/src/github.com/influxdata/telegraf/plugins/inputs/docker_log/docker_logs.go:276 +0x394
github.com/influxdata/telegraf/plugins/inputs/docker_log.(*DockerLogs).tailContainerLogs(0xc00048a380, 0xc000309400, 0x40, 0xc0007160c0, 0x1, 0x4, 0xc0003a02d0, 0x23, 0xc00030a5f0, 0x47, ...)
        /home/dbn/src/influx/src/github.com/influxdata/telegraf/plugins/inputs/docker_log/docker_logs.go:257 +0x5a1
github.com/influxdata/telegraf/plugins/inputs/docker_log.(*DockerLogs).containerListUpdate.func1(0xc00048a380, 0x2a03680, 0xc000140640, 0xc000309400, 0x40, 0xc0007160c0, 0x1, 0x4, 0xc0003a02d0, 0x23, ...)
        /home/dbn/src/influx/src/github.com/influxdata/telegraf/plugins/inputs/docker_log/docker_logs.go:209 +0x220
created by github.com/influxdata/telegraf/plugins/inputs/docker_log.(*DockerLogs).containerListUpdate
        /home/dbn/src/influx/src/github.com/influxdata/telegraf/plugins/inputs/docker_log/docker_logs.go:192 +0x2eb
```